### PR TITLE
Keyboard essentials

### LINKED ISSUE #45: Maturation
Mature the source and work on individual units as well.

Mature the project, add development goals, milestones, organize development and design.

Also, glad to see the project coming back. I've been pretty bored recently so if you'd like another dev, you've got one.

### LINKED ISSUE #54: Keyboard driver
Specs coming soon

Header file: `kernel/include/kernel/keyboard.h`
Source file: `kernel/arch/x86/drivers/keyboard.c`

### LINKED ISSUE #56: Unhandled exception upon enabling interrupts
Upon `enable();` in kernel.c, we get an unhandled exception. Interrupt 32 is raised, which is the correct interrupt for IRQ 0, but the handler called is the one in idt.c, which is not the one that should be called. The issue is probably somewhere in wrappers.asm, irq.c, and/or isr.c.

### PR DESCRIPTION
**Pull request by BrandonKoerner**
ADD TO CHANGELOG:
- created scancode table 1 based on keyboard scancodes for 8042 PS/2 legacy keyboard scheme
- created conditional compilation in kernel/arch/x86/drivers/keyboard.c's `keyboard_install()` which may set the keyboard driver's handler as a callback for IRQ1
- created `__keystate__` to store the bitmap of currently pressed keys
- created `__keystate_ext__` in accordance with extended scancodes for the 8042 PS/2 keyboard
- created `__keyerror__` to store additional information about the keyboard state
- created `keystate_table` and `keystate_ext_table` for scancode to keycode lookups
- created `keyboard_state` struct in kernel/incude/kernel/keyboard.h
- created `keyboard_handler` which updates keyboard state variables accordingly

PLANNED FEATURES:
- use struct `keyboard_state` to store more abstracted information about keyboard state
- implement a circular queue to store keycodes

### CODE REVIEW

**Comment by BrandonKoerner** [comment]
ADDITIONAL INFO TO ADD TO CHANGELOG:
- keyboard handler now calls auxiliary update method `update_keyboard_state` which updates `keyboard_state` and takes additional action, depending on which `keyboard_mode` the keyboard driver is in, on the circular queue.
- keyboard driver supports six modes of operation:
1. `KEYALL`: specifies that queue buffer will store keycodes and will halt driver on queue overflow
2. `KEYDUMP`: specifies that queue buffer will store keycodes and will ignore incoming keycodes when queue buffer is full
3. `KEYSMASH`: specifies that queue buffer will store keycodes and will dequeue keycodes without further action to make room for incoming keycodes as needed
4. `TEXTALL`: specifies that queue buffer will store text and will halt driver on queue overflow
5. `TEXTDUMP`: specifies that queue buffer will store text and will ignore incoming text when queue buffer is full
6. `TEXTSMASH`: specifies that queue buffer will store text and will dequeue text without further action to make room for incoming text as needed
- in `keyboard.h`, provides method `keyboard_install` which requires a `keyboard_mode` argument, a buffer argument, and a buffer size argument. Latter two arguments and space inside the buffer are for sole use by the keyboard driver's queue.
- also in `keyboard.h` is a stateful function for converting keycodes to text: `keycode_to_text`
- keycodes or text data in the keyboard driver's queue is retrieved by `next_keycode` (`keyboard.h`)
- added code in `kernel.c` for testing keyboard driver
- added code in `pit.c` for testing keyboard driver (conditional, can be ignored by removing definition of KEYBOARD_DEBUG in `keyboard.h`)
- in `terminal.c`, changed TERMINAL_HEIGHT to reflect the default height used by qemu emulator
- in `terminal.c`, added backspace support.
- in `terminal.c`, fixed scroll function by utilizing `memcpy` and `memset`

**Review by davidaylaian — CHANGES_REQUESTED** [review]
@ReewassSquared good stuff! I made a cleanup commit at 8c79e4e1ebd167d47f6644e380021993230aea11. Before I rebase and merge, just one more thing: can you `#define` some of the magic numbers used in `keycode_to_text` and `keyboard_handler`?

File: kernel/kernel/arch/x86/drivers/keyboard.c [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+                _r += 0x20;
+            else if ((_r > 0x60) && (_r < 0x7B))
+                _r -= 0x20;
+        }
+
+        return _r;
+    }
+
+    return '\0';
+}
```
Comment (davidaylaian) [inline kernel/kernel/arch/x86/drivers/keyboard.c]: Lots of magic numbers here

File: kernel/kernel/arch/x86/drivers/keyboard.c [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+
+    keyboard_state.lastkey = keycode;
+}
+
+uint8_t next_keycode()
+{
+    return dequeue();
+}
+
+void keyboard_handler()
```
Comment (davidaylaian) [inline kernel/kernel/arch/x86/drivers/keyboard.c:281]: Lots of magic numbers here also

**Comment by BrandonKoerner** [comment]
Will de-mistify code by adding descriptive preprocessor macros for magic numbers

for `keycode_to_text`, the numbers 0x40 and 0x5B (and 0x60 and 0x7B) are the range of alphabetic uppercase and alphabetic lowercase ASCII values, and 0x20 is the constant distance between ASCII alphabetic uppercase and their respective lowercase, and vice versa, which I think doesn't merit magicification.

**Comment by davidaylaian** [comment]
Could those numbers instead be written as expressions of `'A'`, `'Z'`, `'a'`, and `'z'`? Or, even better: could some of the code be replaced by calls to ctype.h?

**Review by BrandonKoerner — COMMENTED** [review]
yes

**Review by davidaylaian — APPROVED** [review]

**Pull request merged**